File: 1232-check-if-it-is-a-straight-line/1232-check-if-it-is-a-straight-line.py
class Solution:
    def checkStraightLine(self, coordinates: List[List[int]]) -> bool:
        if len(coordinates)==2:
            return True
        n = len(coordinates) 
        point1=coordinates[0]
        point2=coordinates[1]
        def checkAllEqual(pos, val):
            for i in range(2, n):
                if coordinates[i][pos] != val:
                    return False
            return True
        
        def checkRatio(px,py, ratio):
            for i in range(2, n):
                
                qx,qy= coordinates[i] 
                print(px, py, qx,qy)
                if py-qy==0:
                    return False
                if (px-qx)/(py-qy) != ratio:
                    return False
            return True
        # veritcal line
        if point1[0] == point2[0]: 
            return checkAllEqual(0, point1[0])
        
        # horizontal line
        if point1[1] == point2[1]:
            return checkAllEqual(1, point1[1])
        

        slope = (point1[0]-point2[0])/(point1[1]-point2[1])
        return checkRatio(point1[0], point1[1], slope) 

File: 1232-check-if-it-is-a-straight-line/1232-check-if-it-is-a-straight-line.js
/**
 * @param {number[][]} coordinates
 * @return {boolean}
 */
var checkStraightLine = function(coordinates) {
    if(coordinates.length===2) return true
    let point1=coordinates.shift()
    let point2=coordinates.shift()
    function checkAllEqual(pos, val){
        for(let c of coordinates){
            if(c[pos]!==val)
                return false
        }
        return true
    }
    function checkRatio(px,py, ratio){
        for(let [qx,qy] of coordinates){
            if((px-qx)/(py-qy)!==ratio)
                return false
        }
        return true
    }
    if(point1[0]===point2[0]){ //veritcal line
        return checkAllEqual(0, point1[0])
    }
    if(point1[1]===point2[1]){ //horizontal line
        return checkAllEqual(1, point1[1])
    }
   
    let ratio=(point1[0]-point2[0])/(point1[1]-point2[1])
    return checkRatio(...point1, ratio) 
    
};